src/utils/validators.js

```diff
@@ -0,0 +1,65 @@
+import { ValidationError } from './errors.js';
+
+export const validateRequiredFields = (data, fields) => {
+  const missingFields = fields.filter(field => !data[field]);
+  if (missingFields.length > 0) {
+    throw new ValidationError(`Missing required fields: ${missingFields.join(', ')}`);
+  }
+};
+
+export const validateUserId = (userId) => {
+  if (!userId) {
+    throw new ValidationError('User ID is required');
+  }
+  if (typeof userId !== 'string' || userId.trim() ===  '') {
+    throw new ValidationError('user id must be  a non-empty string');
+  }
+};
+
```

Comment:
`validateRequiredFields` and `validateUserId` is okay to be in file src/**utils**/validators.js because it is general validation helpers that can be user for any resources. But `validateListingData` I would better move to some separate file related to the listings entity. For example you can create folder validators and create listing.validator.js file there and put this function there.

Also inside of the `validateListingData` function you can use `validateUserId` and `validateRequiredFields`. And then you can remove it from the controller, and just call a single function `validateListingData()`